### ISSUE TITLE
DayNightCompositor not working in conjunction with crop()

### ISSUE THREAD

**Issue by Plantain** [pytroll/satpy#2059]
**Describe the bug**
At certain times of day, with both GOES16 and Himawari8, the produced imagery is blank.

For GOES16 it seems to be 1940Z as the first timestep that fails.

The key factors seem to be needing a crop, DayNightCompositor, and it seems to happen across a broader range of times using reprojected COG's rather than GeoTIFF's weirdly.

The crop boundary seems significant too, potentially related to the anti-meridian.

**To Reproduce**
    scn.load([composite], generate=False)
    scn = scn.crop(xy_bbox=(-4100000,-5000000,5000000,5000000))
    scn = scn.resample(scn.finest_area(),resampler='native')
    scn.save_dataset(composite, filename=outname, driver='COG', tiling_scheme='GoogleMapsCompatible', compress="JPEG", overview_quality=90, quality=90, num_threads="32", aligned_levels=8)

goes1930.tiff is correct
goes1940.tiff is invalid/blank

http://static2.skysight.io/repro.tgz has a bundle that reproduces the issue, with data.

**Expected behavior**
Imagery should be produced for both times.

**Environment Info:**
 - OS: Ubuntu 20.04
 - Satpy Version: git
 - PyResample Version: git

**Comment by djhoese** [comment]
Dang this makes me nervous. Does this example "fail" consistently? What about this exact data case and the default geotiff driver?

**Comment by djhoese** [comment]
I see your day/night composite is:

```yaml
  true_color_daynight_goes:
    compositor: !!python/name:satpy.composites.DayNightCompositor
    standard_name: true_color_daynight_goes
    lim_low: 90.0
    lim_high: 100.0
    prerequisites:
      - true_color 
      - night_ir_alpha
```

Have you tested any other combinations? Specifically something other than the night_ir_alpha?

**Comment by Plantain** [comment]
This case fails consistently. I am not sure offhand if it fails with
GeoTIFF but I have observed failures with GeoTIFF.

**Comment by Plantain** [comment]
Yes, I experienced failures with a range of combinations all using
DayNightCompositor

**Comment by Plantain** [comment]
I confirmed it fails with simple_image/png and GeoTIFF as well on this case.

**Comment by Plantain** [comment]
It *does* seem to be related to night_ir_alpha after all - changing 'night_ir_alpha' to 'C01' works at least.

I am not able to test night_ir_with_background because for reasons I don't understand, it always fails with:

cannot convert float NaN to integer
The following datasets were not created and may require resampling to be generated: DataID(name='true_color_daynight_goes')
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.8/site-packages/satpy/dataset/data_dict.py", line 169, in __getitem__
    return super(DatasetDict, self).__getitem__(item)
KeyError: 'true_color_daynight_goes'

**Comment by djhoese** [comment]
As mentioned on slack, the night background is a geotiff on a different projection. The `native` resampler does not handle transformation to a new projection, only replicating or aggregating pixels of an existing projection.

I don't remember why that warning/message about NaN shows up, but I think it can be ignored as far as this issue goes. If you want to use that composite you'll need to use the `nearest` (or other) resampler to remap at least the background image to the ABI projection *or* resample all data (ABI and the background) to a new projection (new AreaDefinition). There are a couple different ways to do this most efficiently, but we can talk more about that on slack if you're really curious or if this issue is *only* a problem for the night background composites.

**Comment by djhoese** [comment]
Using your `true_color_daynight_goes` with the data you provided I'm able to make an image for 1940 with my Geo2Grid software using the native resampler.

![image](https://user-images.githubusercontent.com/1828519/159179714-8776ea65-1bc6-4fdf-a714-99e571d513d9.png)

Given what I just said about resampling being needed...I don't know how this was possible.

**Comment by djhoese** [comment]
Ah I see, you had switched it to the night IR with a background image. :man_facepalming: Original statements stand and it seems I am able to make this image composite from your original YAML that you shared for the 19:40 data.

**Comment by Plantain** [comment]
Are you also cropping with Geo2Grid though?

The problem does not occur if don't crop, *or* if I change night_ir to 'C01'.

**Comment by djhoese** [comment]
🤦‍♂️nope forgot to crop. Will try again later

**Comment by djhoese** [comment]
Ok so I tried with cropping and no matter what I do I get an error from the resampling step:

```
ValueError: Expand factor must be a whole number
```

Now I am using satpy `main` where in the last month or two there were small changes to the native resampler *but* I don't think I changed any of the logic for determining how many times pixels should be replicated. So step one on my side is to figure out why we get different results (an actual image versus me getting an exception). The other thing you can check is what you get when, after the crop call, you do:

```python
print([x.shape for x in new_scn])
```

I get the following:

```
[(9980, 9082), (19960, 18164), (19960, 18164), (9980, 9082), (4, 4990, 4541)]
```

These numbers make sense to me (all a factor of 2) but for some reason the resampler decided that the ratio for C01 was 2.173947895791583 and 2.3889011231006387.

This will be fun... :cry:

**Comment by djhoese** [comment]
Ah! I was renaming each of my Scene's to `new_scn` for the cropped scene, and `resampled_scn` for the resampled one. I used your code but forgot to update the area used in the resample call.

**Comment by djhoese** [comment]
If I plot the cos(SZA) before it is linearly scaled in the DayNightCompositor I get:

![coszen](https://user-images.githubusercontent.com/1828519/159196550-7f261bd6-b78c-44b7-bb1c-d244a6a1fb48.png)

If I then let it be scaled based on your limits (90 and 100 degrees) I get:

![clipped_coszen](https://user-images.githubusercontent.com/1828519/159196590-b1d6ab64-670e-43cc-b91e-2d0815a413de.png)

This is as far as I could get tonight. I'll have to look harder tomorrow.

**Comment by djhoese** [comment]
Ok, I figured it out. At least I think I did. The alpha bands reaches the final `save_dataset` step and performs the default dynamic linear stretch. Put simply, this enhancement will take the min and max of the data and normalize the data based on those limits to 0-1. For the 19:30 (good) case, the Alpha channel has some less than 1 values (partial transparent) so there is a min and a max and the data gets normalized. In the 19:40 (bad) case, the Alpha channel is all 1s (fully opaque) so the min/max end up equaling the same thing which results in the Alpha data being all fills (fully transparent).

Now normally I would say "this is a bug in the linear stretch not handling Alpha channels, it shouldn't touch the Alpha band", but while investigating this I realized that some composites actually depend on the Alpha channel being stretched. For example, the `night_ir_alpha` composite used in this example case uses the 10.3um band as the Alpha channel and scales it using the dynamic linear scaling. So we can't change this behavior, but could have it be configurable. However...

The real solution is that you should define a custom enhancement for your custom composite as I did above. The DayNightCompositor pre-enhances the day and night images that it uses so they should already be on the 0-1 scale that trollimage expects. No additional scaling should be done. It is a big waste of processing time/memory and is likely producing a worse image than it should.

```yaml
 true_color_daynight_goes:
   standard_name: true_color_daynight_goes
   operations: []
```

If someone disagrees then let's discuss. Otherwise, let's close this.
